specs/protocol/safe-extensions-v2.md

```diff
@@ -0,0 +1,489 @@
+
+If this invariant is broken, the event history of the timelock would contain useless events.
+
+## Function Specification
+
+### `configureLivenessModule`
+
+Configure the contract as a liveness module by setting the `liveness_challenge_period` and `fallback_owner`.
+
+- MUST allow an arbitrary number of `safe` contracts to use the contract as a module.
+- The contract MUST be enabled as a module on the `safe`.
+- MUST set the caller as a `safe`.
+- MUST take `liveness_challenge_period` and `fallback_owner` as parameters and store them as related to the `safe`.
+- If a challenge exists, it MUST be canceled, including emitting the appropriate events.
+- MUST emit a `ModuleConfigured` event with at least `liveness_challenge_period` and `fallback_owner` as parameters.
+
+### `clearLivenessModule`
+
+Removes the liveness module configuration by a previously enabled `safe`.
+
+- The contract MUST NOT be enabled as a module on the `safe`.
+- MUST erase the existing `liveness_challenge_period` and `fallback_owner` data related to the calling `safe`.
+- If a challenge exists, it MUST be cancelled, including emitting the appropriate events.
+- MUST emit a `ModuleCleared` event.
+
+### `configureTimelockGuard`
+
+Configure the contract as a timelock guard by setting the `timelock_delay`.
+
+- MUST allow an arbitrary number of `safe` contracts to use the contract as a guard.
+- The contract MUST be enabled as a guard for the `safe`.
+- MUST set the caller as a `safe`.
+- MUST take `timelock_delay` as a parameter and store is as related to the `safe`.
+- MUST emit a `GuardConfigured` event with at least `timelock_delay` as a parameter.
+
+### `clearTimelockGuard`
+
+Remove the timelock guard configuration by a previously enabled `safe`.
+
+- The contract MUST NOT be enabled as a guard for the `safe`.
+- MUST erase the existing `timelock_delay` data related to the calling `safe`.
+- If a challenge exists, it MUST be cancelled, including emitting the appropriate events.
+- MUST emit a `GuardCleared` event.
+
+### `viewLivenessModuleConfiguration`
+
+Returns the `liveness_challenge_period` and `fallback_owner` for a given `safe`.
+
+- MUST never revert.
+
+### `viewTimelockGuardConfiguration`
+
+Returns the `timelock_delay` for a given `safe`.
+
+- MUST never revert.
+
+### `getLivenessChallengePeriodEnd`
+
+Returns `challenge_start_time + liveness_challenge_period + timelock_delay` if there is a challenge for the given
+`safe`, or 0 if not.
+
+- MUST never revert.
+
+### `challenge`
+
+Challenges an enabled `safe` to prove that it is not in a liveness failure state.
+
+- MUST only be executable by `fallback` owner of the challenged `safe`.
+- MUST revert if the `safe` hasn't enabled the contract as a module.
+- MUST revert if the `safe` hasn't configured the contract as a module.
+- MUST revert if a challenge for the `safe` exists.
+- MUST set `challenge_start_time` to the current block time.
+- MUST emit the `ChallengeStarted` event.
+
+### `respond`
```

Comment:
You could put the `respond` logic in `checkAfterExecution`, which would make that any successful transaction by the safe clears the liveness challenge, but then you would be forced to install the contract as a module and a guard for liveness.

Right now, if we don't want to enable the timelock, we just don't enable the contract as a guard. If we would want to use `checkAfterExecution` then we would have to always enable the contract as a module and as a guard, and if the `timelock_delay` is 0, then `checkTransaction` would always suceed.

Reply:
I prefer to have them separated